### ISSUE TITLE
High-performance read operations (especially for large numbers of files)

### ISSUE THREAD

**Issue by josephrocca** [WICG/native-file-system#57]
I wanted to ask about @pwnall's comment on another issue (#56) but figured it would be off topic on that issue, so here's part of that comment:

> This API's mission is to facilitate exchanging data with native applications. I think that performance should be a secondary concern. For high-performance I/O, I recommend using sandboxed APIs like Cache Storage or IndexedDB.

My particular use case for this API: Currently for my application the user needs to select a folder of images (can be tens of thousands of images) and then the application processes these images and shows the user a visualisation. The user can tweak some parameters and then re-run the processing to get a new visualisation.

Some benefits of this new API for my use-case would be:

 * Allowing the user to select ~~multiple files (or a whole folder)~~ **a directory** in a cross-browser way. The non-standard ~~`multiple`~~ `webkitdirectory` attribute on file inputs is [only supported by Chrome](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/webkitdirectory).
 * The user doesn't have to keep selecting their image folder every time they load the application - they just select it once at the start and then application will "remember" a reference to that folder when they next return (they will tend to return regularly - sometimes many times per day).

But I actually thought the main benefit would be higher-performance, non-blocking file-reading. Selection of a folder of 10,000 images using the `multiple` files input on Chrome causes several seconds of lag on the main thread (this is before actually reading the files at all). So the `<input type="file">` approach doesn't seem to be designed for a large number of files, and so even if it was cross-browser, it doesn't seem like an ideal approach. Though maybe the main-thread blocking is just an implementation-specific thing - but even so, it doesn't seem like it should take 5+ seconds (on a decent laptop) to get a reference to 10k files.

tl;dr is that my use case requires high-performance reading (and ideally writing too), so I'm hoping that the API design takes into account this sort of use-case. I'd have thought that performance would be a top priority if the API intends to bring near-native file-wrangling capabilities to the web. But I must admit I've only just discovered this API and skimmed the spec, so maybe I'm imagining that it's something that it's not.

**Comment by guest271314** [comment]
> - Allowing the user to select multiple files (or a whole folder) in a cross-browser way. The non-standard `multiple` attribute on file inputs is only supported by Chrome.

`multiple` attribute at `<input type="file">` is supported at  Chrome, Edge, Firefox, Opera, Safari https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/multiple#Browser_compatibility.  `webkitdirectory` and `allowdirs` can also be used at Chrome and Firefox respectively to allow directories to be uploaded to `<input type="file">` element (https://stackoverflow.com/q/39664662).

> But I actually thought the main benefit would be higher-performance, non-blocking file-reading. Selection of a folder of 10,000 images using the multiple files input on Chrome causes several seconds of lag on the main thread (this is before actually reading the files at all). So the <input type="file"> approach doesn't seem to be designed for a large number of files, and so even if it was cross-browser, it doesn't seem like an ideal approach. Though maybe the main-thread blocking is just an implementation-specific thing - but even so, it doesn't seem like it should take 5+ seconds (on a decent laptop) to get a reference to 10k files.

What is available RAM and disk space at the different machines the code was tried? The time different machines can take to process data (using the same code) can varying considerably. For example, at a machine having minimal RAM and available disk space the fetch and download of a 189MB took 4:20 (4 minutes and 20 seconds), the same code at a different machine took seconds (https://stackoverflow.com/q/39959467). In general, more procedures takes more time (https://stackoverflow.com/a/47173390).

**Comment by josephrocca** [comment]
Sorry, I meant `webkitdirectory`, not `multiple`. Getting the user to select 50,000 images by navigating into the folder and multi-selecting (by some means, depending on the device) is obviously absurd. They should just be able to select the directory that contains the images.

I don't think your RAM/disk-space comments are relevant to this issue because the lag occurs before I even use `FileReader` to read the files (I obviously don't read them into memory all at once). The lag occurs just due to actually selecting the directory containing the files. To give you an idea, an a decently fast laptop with 16GB of RAM, it lags the main thread for more than 1 minute when loading a folder that contains about 200,000 images. The lag seems to scale linearly with the number of images. But in any case, this is sort of heading off-topic.

**Comment by guest271314** [comment]
> Getting the user to select 50,000 images by navigating into the folder and multi-selecting (by some means, depending on the device) is obviously absurd.

It is currently possible to select or drop a folder at Chromium/Chrome or Firefox using `webkitdirectory` and `allowdirs` without selecting the files and/or directories within the folder, and then process each sub-directory and file. See the code linked at previous comment. 

> I don't think your RAM/disk-space comments are relevant to this issue because the lag occurs before I even use FileReader to read the files (I obviously don't read them into memory all at once).

The uploaded files still need to be read (see https://www.w3.org/TR/FileAPI/#snapshot-state; Where is Blob binary data stored? https://stackoverflow.com/q/38239361). RAM and available disk space and possibly other running programs do effect output (time).

For reading `File` objects as a stream see https://chromium-review.googlesource.com/c/chromium/src/+/1526796.

It is not clear what you are expecting? What is the reason you believe the reading of 200,000 files should  take than 1 minute? Have you tried the same code at different machines?

**Comment by guest271314** [comment]
The user can also currently select one or more directories from native file manager GUI and drop the folders at a `<input type="file">` element without selecting _N_ individual files.

**Comment by josephrocca** [comment]
The directory attributes are non-standard and not cross-browser. Again, this is not about reading the files. It's simply about getting a reference to the directory so that they can be accessed on `inputElement.files[n]` and later read (which isn't done one the main thread, and thus doesn't cause lag). Simply getting a reference is what's causing the lag. To re-iterate: This has nothing to do with reading the contents of the files. In any case, I appreciate the effort here, but I'm not asking for help with my application - that's off topic.

This issue is in regard to the direction of this proposal - specifically the priority being placed on performance. I thing that performance should be a primary concern for an API like this, and reading the below comment by @pwnall makes me concerned that we're going to end up with an API that doesn't cater to the needs of many of the most interesting use cases that API enables.

> This API's mission is to facilitate exchanging data with native applications. I think that performance should be a secondary concern. For high-performance I/O, I recommend using sandboxed APIs like Cache Storage or IndexedDB.

**Comment by guest271314** [comment]
> The directory attributes are non-standard.

Using the appropriate code they work for allowing and processing directories at Firefox and Chromium/Chrome.

> Again, this is not about reading the files.

The uploaded folder and/or files need to be read by the browser to get a reference, else what underlying data would the reference be pointing to? The upload of files by a user is not an _instantaneous_ transfer of files from the user filesystem, potentially over a network connection, to the browser configuration folder or RAM that requires zero time and space. 

> so that they can be accessed on `inputElement.files[n]` and later read (which isn't done one the main thread, and thus doesn't cause lag)

By read meaning the process described at the linked questions and answers (https://github.com/WICG/native-file-system/issues/57#issuecomment-498057269). Each file is read into memory and/or written to the user filesystem (e.g., Chromium/Chrome configuration folder) - that is before any user defined read using `FileReader`, `Response.arrayBuffer()`, etc. 

How do you propose for this API to read 200,000 files faster than the files are currently read at upload to `<input type="file">` element?

What is the MIME type and average size (in bytes) of the 200,000  image files?

**Comment by mkruisselbrink** [comment]
One problem with the current chrome implementation of `<input type=file webkitdirectory>` (which despite the name is actually implemented by all major browsers afaik) is that it (recursively) enumerates all the files in the selected directory (i.e. https://wicg.github.io/entries-api/#issue-452f97f6). It doesn't have to do that, since it could just return a single directory entry rather than individual files for all the files in the directory, but I don't think we're likely to change the implementation there, for risk of breaking compatibility with existing websites. For this API on the other hand we definitely want to return just a single directory handle rather than directly returning handles for all the contents of that directory, so this should be somewhat/a lot more performant in that regard.

**Comment by mkruisselbrink** [comment]
Closing this issue, since I believe the API as designed already addresses these concerns.
